### ISSUE TITLE
Support assignment to union of `TypedDict`

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#2265]
We currently only consider the type context if there is a single `TypedDict` element. We should support the following:

```py
from typing import TypedDict

class Foo(TypedDict):
    foo: int

class Bar(TypedDict):
    bar: int

x: Foo | Bar = reveal_type({ "foo": 1 }) # Foo

class FooBar1(TypedDict):
    foo: int
    bar: int

class FooBar2(TypedDict):
    foo: int
    bar: int

x: FooBar1 | FooBar2 = reveal_type({ "foo": 1, "bar": 1 }) # FooBar1 | FooBar2
```

**Comment by ibraheemdev** [comment]
Ah, interestingly, the examples above type check, but fail if I remove the `reveal_type` call, because we attempt to narrow the type context of the generic call.
